bytecomp/matching.ml

```diff
@@ -1562,59 +1672,75 @@ let inline_lazy_force arg loc =
          tables (~ 250 elts); conditionals are better *)
       inline_lazy_force_cond arg loc
 
-let make_lazy_matching def = function
+let make_lazy_matching loc def = function
     [] -> fatal_error "Matching.make_lazy_matching"
-  | (arg,_mut) :: argl ->
+  | (arg, _mut, _arg_loc) :: argl ->
       { cases = [];
         args =
-          (inline_lazy_force arg Location.none, Strict) :: argl;
+          (inline_lazy_force arg loc, Strict, loc) :: argl;
         default = make_default matcher_lazy def }
 
 let divide_lazy p ctx pm =
   divide_line
     (filter_ctx p)
-    make_lazy_matching
+    (make_lazy_matching p.pat_loc)
     get_arg_lazy
     p ctx pm
 
 (* Matching against a tuple pattern *)
 
 
 let get_args_tuple arity p rem = match p with
-| {pat_desc = Tpat_any} -> omegas arity @ rem
+| {pat_desc = Tpat_any} -> omegas p.pat_loc arity @ rem
 | {pat_desc = Tpat_tuple args} ->
     args @ rem
 | _ ->  assert false
 
 let matcher_tuple arity p rem = match p.pat_desc with
 | Tpat_or (_,_,_)     -> raise OrPat
 | Tpat_any
-| Tpat_var _ -> omegas arity @ rem
+| Tpat_var _ -> omegas p.pat_loc arity @ rem
 | Tpat_tuple args when List.length args = arity -> args @ rem
 | _ ->  raise NoMatch
 
-let make_tuple_matching loc arity def = function
+let make_tuple_matching arity field_pat_locs def = function
     [] -> fatal_error "Matching.make_tuple_matching"
-  | (arg, _mut) :: argl ->
+  | (arg, _mut, _arg_loc) :: argl ->
       let rec make_args pos =
         if pos >= arity
         then argl
-        else (Lprim(Pfield pos, [arg], loc), Alias) :: make_args (pos + 1) in
+        else begin
+          assert (pos >= 0 && pos < Array.length field_pat_locs);
```

Comment:
Now that you get passed a list/array of the same size as the output, you should use `List.mapi` instead of `make_args`, the code would be simpler. (I think that `field_pat_locs` could be passed as a list, which more coherent with the rest of the code.)

Reply:
Likewise